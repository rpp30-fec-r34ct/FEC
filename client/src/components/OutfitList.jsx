import React from 'react';
import AddCard from './AddCard.jsx';
import OutfitCard from './OutfitCard.jsx';


export default function OutfitList(props) {
  return (
    <div className="outfit-list">
      <AddCard />
      {/* <OutfitCard /> */}
    </div>
  )
}