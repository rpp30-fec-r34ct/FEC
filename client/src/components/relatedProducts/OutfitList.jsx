import React, { useState, useEffect } from 'react'
import AddCard from './AddCard.jsx'
import OutfitCard from './OutfitCard.jsx'
import { AiOutlinePlusCircle } from 'react-icons/ai'
import { FaChevronLeft, FaChevronRight } from 'react-icons/fa'
import { useParams } from 'react-router-dom'

export default function OutfitList({currentProduct}) {
  const [userOutfits, setUserOutfits] = useState([])
  const [outfitPosition, setOutfitPosition] = useState(0)
  const [outfitIndex, setOutfitIndex] = useState(0)


  useEffect(() => {
    setUserOutfits(JSON.parse(window.localStorage.getItem('outfit')))
  }, [])

  //Edit the local storage array ONLY the outfit list changes using dependency array
  useEffect(() => {
    window.localStorage.setItem('outfit', JSON.stringify(userOutfits))
  }, [userOutfits])


  const addOutfit = () => {
    const temp = JSON.parse(window.localStorage.getItem('outfit')) || []

    if (!temp.find((outfit) => outfit.id === currentProduct.id)) {
      temp.push(currentProduct)
      window.localStorage.setItem('outfit', JSON.stringify(temp))
    }
    setUserOutfits(temp)
  }

  const deleteOutfit = (selectedId) => {
    const oldTemp = JSON.parse(window.localStorage.getItem('outfit'))
    const newTemp = oldTemp.filter(outfit => outfit.id !== selectedId)

    setUserOutfits(newTemp)
  }


  const nextOutfit = () => {
    setOutfitIndex(outfitIndex => outfitIndex + 1)
    setOutfitPosition(outfitPosition - 220)
  }

  const prevOutfit = () => {
    setOutfitIndex(outfitIndex => outfitIndex - 1)
    setOutfitPosition(outfitPosition + 220)
  }

  return (
    <div className='outfit-overview'>
      <h3>YOUR OUTFIT</h3>
        <div className='carousel-container'>
            <AddCard addOutfit={addOutfit} />
          {outfitPosition < 0 && <FaChevronLeft className='left-arrow' onClick={prevOutfit} />}
            <div className='carousel-content-wrapper'>
              <div className='carousel-content' style={{ transform: `translateX(-${outfitIndex * 25}%)` }}>
                {
                  userOutfits.map((outfit) => <OutfitCard key={outfit.id} outfit={outfit} deleteOutfit={deleteOutfit}/>)
                }
        </div>
       </div>
      {userOutfits.length > 3 && outfitIndex < (userOutfits.length - 3) && <FaChevronRight className='right-arrow' onClick={nextOutfit} />}
    </div>
  </div>
  )
}
