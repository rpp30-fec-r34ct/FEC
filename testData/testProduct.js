/* eslint-disable */
const testProduct = {
  id: 47421,
  campus: 'hr-rpp',
  name: 'Camo Onesie',
  slogan: 'Blend in to your crowd',
  description: 'The So Fatigues will wake you up and fit you in. This high energy camo will have you blending in to even the wildest surroundings.',
  category: 'Jackets',
  default_price: '140.00',
  created_at: '2021-08-26T20:30:48.129Z',
  updated_at: '2021-08-26T20:30:48.129Z',
  features: [
    {
      feature: 'Fabric',
      value: 'Canvas'
    },
    {
      feature: 'Buttons',
      value: 'Brass'
    }
  ]
}

export default testProduct