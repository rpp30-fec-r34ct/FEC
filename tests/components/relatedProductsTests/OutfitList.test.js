
/* eslint-disable */
/**
 * @jest-environment jsdom
 */
import React from 'react'
import { render } from '@testing-library/react'
import 'regenerator-runtime/runtime'
import '@testing-library/jest-dom'
import { createMemoryHistory } from 'history'
import { Router, Route, Switch } from 'react-router-dom'
import testProduct from '../../../testData/testProduct'
import OutfitList from '../../../client/src/components/relatedProducts/OutfitList.jsx'

let app


let windowSpy;

// beforeEach(() => {
//   windowSpy = jest.spyOn(window, "window", "get");
// });

// afterEach(() => {
//   windowSpy.mockRestore();
// });

// it('should return https://example.com', () => {
  // windowSpy.mockImplementation(() => ({
  //   location: {
  //     origin: "https://example.com"
  //   }
  // }));

//   expect(window.location.origin).toEqual("https://example.com");
// });

// it('should be undefined.', () => {
//   windowSpy.mockImplementation(() => undefined);

//   expect(window).toBeUndefined();
// });

var localStorageMock = (function() {
    var store = {};

    return {
        getItem: function(key) {
            return store[key] || null;
        },
        setItem: function(key, value) {
            store[key] = value.toString()
        },
        clear: function() {
            store = {};
        }
    };
})();

Object.defineProperty(global, 'localStorage', {
  value: localStorageMock
})


describe('Outfit List Component', function () {
  beforeEach(() => {
    Object.defineProperty(window, "localStorage", {
      // value: localStorageMock,
      value: {
        getItem: jest.fn(() => null),
        setItem: jest.fn(() => null)
      },
      writable: true
    });

    const history = createMemoryHistory()
    const route = '/product/47421/carousel'
    history.push(route)
    app = render(
      <Router history={history}>
        <Switch>
          <Route path='/product/:productId/carousel/'>
            <OutfitList currentOverview={testProduct} />
          </Route>
        </Switch>
      </Router>
    )

  })
  test('Outfit List should render', () => {
    expect(app).toMatchSnapshot()
  })
  test('Outfit List should initially have no items', () => {
    expect(app.container.querySelector('.carousel-content').children.length).toEqual(0)
  })
  test('Outfit List should have a header', () => {
    expect(app.container.querySelector('.outfit-list-header-1').innerHTML).toBe('YOUR OUTFIT')
  })

  test('Outfit List should get item from local storage', () => {
      expect(window.localStorage.getItem).toHaveBeenCalledTimes(1)
  })
})
