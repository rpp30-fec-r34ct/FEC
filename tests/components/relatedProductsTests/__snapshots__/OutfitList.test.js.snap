// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outfit List Component Outfit List should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h3
        class="outfit-list-header-1"
        style="margin-top: 5px;"
      >
        YOUR OUTFIT
      </h3>
      <div
        class="outfit-list"
      >
        <div
          class="carousels-overview"
        >
          <div
            class="carousel-container"
            style="width: 870px;"
          >
            <div
              class="add-card"
              style="width: 100%; max-width: 200px;"
            >
              <div
                class="add-details"
              >
                <div
                  class="add-btn"
                  data-testid="add-outfit-button"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 1024 1024"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
                    />
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                  </svg>
                </div>
                Add to Outfit
              </div>
            </div>
            <div
              class="carousel-content-wrapper"
            >
              <div
                class="carousel-content"
                style="transform: translateX(-0%);"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <h3
      class="outfit-list-header-1"
      style="margin-top: 5px;"
    >
      YOUR OUTFIT
    </h3>
    <div
      class="outfit-list"
    >
      <div
        class="carousels-overview"
      >
        <div
          class="carousel-container"
          style="width: 870px;"
        >
          <div
            class="add-card"
            style="width: 100%; max-width: 200px;"
          >
            <div
              class="add-details"
            >
              <div
                class="add-btn"
                data-testid="add-outfit-button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 1024 1024"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
                  />
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                </svg>
              </div>
              Add to Outfit
            </div>
          </div>
          <div
            class="carousel-content-wrapper"
          >
            <div
              class="carousel-content"
              style="transform: translateX(-0%);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
